### DISCUSSION TITLE
Unexpected behavior with RavenDB client for NodeJS

### DISCUSSION THREAD

**Discussion by AnjanaRatnayake** [ravendb/ravendb#15342] [Q&A]
I am running a RavenDB cluster of 3 nodes:
```
const raven1             = 'http://raven1:8080';                        // Leader node
const raven2             = 'http://raven2:8080';                        // Follower node
const raven3             = 'http://raven3:8080';                        // Follower node
const server_urls         = [raven1, raven2, raven3];
const database_name       = 'Bank';
```
**Code block 1**
I first add data to the 'Bank' database as follows:
```
let customer = {
    name: "foo",
    balance: 10
}

{
    // Here we insert a customer into the Banks database
    let documentstore = new DocumentStore(server_urls[0], database_name);
    await documentstore.initialize();

    let session = await documentstore.openSession();
    await session.store(customer)
    session.advanced.useOptimisticConcurrency = true;
    await session.saveChanges();

}
```
**Code block 2**
When I run the code below I get the expected behavior from the database. On ```session.saveChanges``` the first session throws a ```ConcurrencyException``` and the change to the balance variable from the first session is not persisted to the database and the change from the second session is persisted.
```
let documentstore = new DocumentStore(server_urls[1], database_name);
await documentstore.initialize();

{

    let session = await documentstore.openSession();
    session.advanced.useOptimisticConcurrency = true;
    
    // Here we load the same customer twice and change the balance
    let customer = await session.load(customer.id);
    let customer_2 = await session.load(customer.id);

    customer.balance += 100;
    customer_2.balance += 100;
    
    {
        // Before saving the outer session we open another session and change the balance a different way.
        // With optimistic concurrency turned on we would expect the outer session to now be invalid.
        let session_cpy =  await documentstore.openSession({
            transactionMode: "ClusterWide"
        });

        let customer_cpy = await session_cpy.load(customer.id);
        let customer_cpy_2 = await session_cpy.load(customer.id);

        customer_cpy.balance += 100;
        customer_cpy_2.balance -= 100;

        await session_cpy.saveChanges();
        
    }
    try{
        await session.saveChanges();
    } catch(e) {
        console.log(e);
    }
```

I noticed that when I changed the way I instantiated the data in **Code block 1** to the code below - the ```ConcurrencyException``` is no longer triggered in **Code block 2** and both changes are persisted to the database.

```
// let customer = {
//     name: "foo",
//     balance: 10
// }

class Customer {
    constructor() {
        this.name = "foo"
        this.balance = 10;
    }
}

let customer = new Customer();
{
    // Here we insert a customer into the Banks database
    let documentstore = new DocumentStore(server_urls[0], database_name);
    await documentstore.initialize();

    let session = await documentstore.openSession();
    await session.store(customer)
    session.advanced.useOptimisticConcurrency = true;
    await session.saveChanges();

}
```
 I am not sure why optimistic concurrency behavior changes depending on whether the data is instantiated as a Class or an Object. Not sure if this is expected behavior or a bug?

**Comment by ayende** [comment]
Note that passing `server_urls[0]` or `server_urls[1]` doesn't actually *change* which server we use.
RavenDB discover the topology on the cluster and will use the right node for this based on the database configuration.

Note that you are also mixing cluster wide and non cluster wide transactions.

What does the _document_ looks like in both cases? 

Are you using distinct ids on both cases?

There should be no difference between the options as far as the server is concerned.

**Reply by AnjanaRatnayake** [reply]
Yes I am using distinct ids in both cases. Here is an example of what the document looks like when a Class is used:
```
Customer { name: 'foo', balance: 10, id: 'customers/129-A' }
```
Here is an example of what a document looks like when an Object is used:
```
{ name: 'foo',
  balance: 10,
  id: 'd4ebba96-2118-4ff8-9d82-75a7547a4828' }
```
They are distinct IDs. 
This is the code I ran to obtain this:
```
(async () => {
    const { DocumentStore }  = require('ravendb');
    const raven1             = 'http://raven1:8080';
    const raven2             = 'http://raven2:8080';
    const raven3             = 'http://raven3:8080';
    const server_urls         = [raven1, raven2, raven3];
    const database_name       = 'Bank';
    
    let customer = {
        name: "foo",
        balance: 10
    }
    
    // class Customer {
    //     constructor() {
    //         this.name = "foo"
    //         this.balance = 10;
    //     }
    // }
    
    // let customer = new Customer();
    await insert_data(server_urls, database_name, customer, DocumentStore);
    await edit_data(server_urls, database_name, customer, DocumentStore);
})();

async function insert_data(server_urls, database_name, customer, DocumentStore){
    let documentstore = new DocumentStore(server_urls[0], database_name);
    await documentstore.initialize();

    let session = await documentstore.openSession();
    await session.store(customer)
    session.advanced.useOptimisticConcurrency = true;
    await session.saveChanges();
    console.log(customer);
}

async function edit_data(server_urls, database_name, customer, DocumentStore){
    let documentstore = new DocumentStore(server_urls[1], database_name);
    await documentstore.initialize();    
    let session = await documentstore.openSession();
    session.advanced.useOptimisticConcurrency = true;
    
    // Here we load the same customer twice and change the balance
    customer = await session.load(customer.id);
    let customer_2 = await session.load(customer.id);

    customer.balance += 100;
    customer_2.balance += 100;
    // Before saving the outer session we open another session and change the balance a different way.
    // With optimistic concurrency turned on we would expect the outer session to now be invalid.
    let session_cpy =  await documentstore.openSession({
        transactionMode: "ClusterWide"
    });

    let customer_cpy = await session_cpy.load(customer.id);
    let customer_cpy_2 = await session_cpy.load(customer.id);

    customer_cpy.balance += 100;
    customer_cpy_2.balance -= 100;

    await session_cpy.saveChanges();
    
    await session.saveChanges();

}
```
I am still only getting ConcurrencyException when the Object is used and not when the Class is used. It is weird that the behavior of this code is dependent on whether I use 
```
    class Customer {
        constructor() {
            this.name = "foo"
            this.balance = 10;
        }
    }
    
    let customer = new Customer();
```
OR
```
    let customer = {
        name: "foo",
        balance: 10
    }
```
as the document I want to store

**Reply by ayende** [reply]
This is a bug here: 
https://github.com/ravendb/ravendb-nodejs-client/blob/b2b8e4b0e1f06bd0d32a24bb3ac5ba3ce78c6cc6/src/Documents/Session/InMemoryDocumentSessionOperations.ts#L1779-L1780

The problem is that if you are using a raw object, we read the `Raven-Node-Type` (which is `null`, in this case).
But `null` is assumed to be a dirty value, so we force an spurious update. When you have a type, this works properly.

**Reply by ayende** [reply]
https://issues.hibernatingrhinos.com/issue/RDBC-653/Raw-object-literal-in-node.js-is-always-considered-dirty

**Reply by AnjanaRatnayake** [reply]
> This is a bug here: https://github.com/ravendb/ravendb-nodejs-client/blob/b2b8e4b0e1f06bd0d32a24bb3ac5ba3ce78c6cc6/src/Documents/Session/InMemoryDocumentSessionOperations.ts#L1779-L1780
> 
> The problem is that if you are using a raw object, we read the `Raven-Node-Type` (which is `null`, in this case). But `null` is assumed to be a dirty value, so we force an spurious update. When you have a type, this works properly.

Maybe I have misunderstood your explanation. What I noticed was that I get the ConcurrencyException when there is no type and I do not get the ConcurrencyException when there is a type. Is this not the opposite of the behavior you described above?

**Reply by ayende** [reply]
Same thing, basically.
You are looking into the behavior, I'm explaining why it is doing this thing
